### PR TITLE
S2 Feature/Story

### PR DESCRIPTION
**Pull request by implex001**
The following features have been added in this pull request. Would love to hear any feedback/ideas.

# Story
- new prologue scene when you start a new game (note pressing escape key will skip the prologue)
- new epilogue scene once you exit the safehouse for the third time

# Story Manager
Singleton class that manages the loading, display and disposal of stories. There are several components that allow the manager to operate.
### StoryBase Interface
A story base object is what the manager calls upon to control the story. It also stores the entities necessary for the story. Current Story Bases are:
- Cutscene
- SimpleDialogue (the sprint 1 dialogue box)
### StoryConfig Abstract Classes and Interface
Stores raw data and methods needed to create the entities. Each story base must have a config.
### StoryNames Enum
IDs of the stories.
### Story Input Component
Adds input to advance the story (Spacebar) and skip the story (ESC) 

# Music
- background music added to the safehouse
- background music added to the game over screen
- background music added for the end boss (currently added to level 3 until the end boss has been created)

### CODE REVIEW

**Comment by implex001** [comment]
While fixing the merge conflict there were two versions of the safehouse music. Which one should be the music for it?
"sounds/safehouse-music.mp3" from feature/story 
OR
"sounds/fireflies-theme-sneak.mp3" from main

For now I will merge the safehouse-music.mp3, but this can be changed later on.

File: source/core/src/main/com/deco2800/game/components/story/StoryBase.java [hunk 1]
```diff
@@ -1,5 +1,30 @@
 package com.deco2800.game.components.story;
```
Comment (25Ericcheong) [inline source/core/src/main/com/deco2800/game/components/story/StoryBase.java:2]: might be worth adding short summary on why we need story base here as well. You could probably just add what you've mentioned in the pull request 'A story base object is what the manager calls upon to control the story. It also stores the entities necessary for the story'

File: source/core/src/main/com/deco2800/game/components/story/StoryConfig.java [hunk 2]
```diff
@@ -0,0 +1,20 @@
+package com.deco2800.game.components.story;
+
```
Comment (25Ericcheong) [inline source/core/src/main/com/deco2800/game/components/story/StoryConfig.java:2]: Similarly for `StoryBase.java` - probably good to have an explanation like you've already written - 'Stores raw data and methods needed to create the entities. Each story base must have a config.'

File: source/core/src/main/com/deco2800/game/components/story/basicdialogue/BasicDialogue.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
+package com.deco2800.game.components.story.basicdialogue;
+
+import com.deco2800.game.components.Component;
+import com.deco2800.game.components.dialoguebox.TextDialogueBox;
+import com.deco2800.game.components.story.StoryBase;
+import com.deco2800.game.entities.Entity;
+import com.deco2800.game.entities.factories.DialogueBoxFactory;
+import com.deco2800.game.services.ServiceLocator;
+
```
Comment (25Ericcheong) [inline source/core/src/main/com/deco2800/game/components/story/basicdialogue/BasicDialogue.java:9]: short explanation on having a `BasicDialogue.java` - where does it come into play and why do you need it maybe.

File: source/core/src/main/com/deco2800/game/components/story/basicdialogue/BasicDialogueConfig.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
+package com.deco2800.game.components.story.basicdialogue;
+
+import com.deco2800.game.components.dialoguebox.Dialogue;
+import com.deco2800.game.components.dialoguebox.DialogueImage;
+import com.deco2800.game.components.story.StoryBase;
+import com.deco2800.game.components.story.StoryConfig;
+
+import java.util.ArrayList;
+import java.util.List;
+
```
Comment (25Ericcheong) [inline source/core/src/main/com/deco2800/game/components/story/basicdialogue/BasicDialogueConfig.java:10]: same for `BasicDialogue.java` - just a short explanation

File: source/core/src/main/com/deco2800/game/components/story/cutscene/CutSceneConfig.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+import com.badlogic.gdx.graphics.Texture;
+import com.badlogic.gdx.scenes.scene2d.ui.Image;
+import com.deco2800.game.components.dialoguebox.Dialogue;
+import com.deco2800.game.components.story.StoryBase;
+import com.deco2800.game.components.story.StoryConfig;
+import com.deco2800.game.services.ServiceLocator;
+
+import java.util.ArrayList;
+import java.util.List;
+
```
Comment (25Ericcheong) [inline source/core/src/main/com/deco2800/game/components/story/cutscene/CutSceneConfig.java:14]: short explanation on why cut scene config is needed perhaps?

File: source/core/src/main/com/deco2800/game/components/story/stories/EpilogueCutScene.java [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    }
+
+    /**
+     * Returns the length of the story
+     *
+     * @return the length of the story
+     */
+    @Override
+    public int getLength() {
+        return 16;
```
Comment (25Ericcheong) [inline source/core/src/main/com/deco2800/game/components/story/stories/EpilogueCutScene.java:70]: could perhaps use a meaningful constant variable

File: source/core/src/main/com/deco2800/game/components/story/stories/IntroDialogue.java [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+    }
+
+    /**
+     * Returns the length of the story
+     *
+     * @return the length of the story
+     */
+    @Override
+    public int getLength() {
+        return 1;
```
Comment (25Ericcheong) [inline source/core/src/main/com/deco2800/game/components/story/stories/IntroDialogue.java:22]: could perhaps use a meaningful constant variable

File: source/core/src/main/com/deco2800/game/components/story/stories/PrologueCutScene.java [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+    }
+
+    /**
+     * Returns the length of the story
+     *
+     * @return the length of the story
+     */
+    @Override
+    public int getLength() {
+        return 25;
```
Comment (25Ericcheong) [inline source/core/src/main/com/deco2800/game/components/story/stories/PrologueCutScene.java:104]: could perhaps use a meaningful constant variable

File: source/core/src/main/com/deco2800/game/components/story/stories/TestCutscene.java [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+        musicPath = musicPathRaw;
+    }
+
+    /**
+     * Returns the length of the story
+     * @return the length of the story
+     */
+    @Override
+    public int getLength() {
+        return 3;
```
Comment (25Ericcheong) [inline source/core/src/main/com/deco2800/game/components/story/stories/TestCutscene.java:28]: could perhaps use a meaningful constant variable

File: source/core/src/main/com/deco2800/game/ui/terminal/commands/StartTestCutSceneCommand.java [hunk 10]
```diff
@@ -0,0 +1,23 @@
+package com.deco2800.game.ui.terminal.commands;
```
Comment (25Ericcheong) [inline source/core/src/main/com/deco2800/game/ui/terminal/commands/StartTestCutSceneCommand.java:1]: was this file used for testing purposes? If yes, will it still be required?

File: source/core/src/main/com/deco2800/game/utils/MenuUtility.java [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+import com.badlogic.gdx.scenes.scene2d.Actor;
+import com.badlogic.gdx.scenes.scene2d.InputEvent;
+import com.badlogic.gdx.scenes.scene2d.ui.*;
+import com.badlogic.gdx.scenes.scene2d.utils.ChangeListener;
+import com.badlogic.gdx.scenes.scene2d.utils.ClickListener;
+import com.deco2800.game.entities.Entity;
+import com.deco2800.game.services.ServiceLocator;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
```
Comment (25Ericcheong) [inline source/core/src/main/com/deco2800/game/utils/MenuUtility.java:13]: short explanation on having menu utility

**Review by 25Ericcheong — APPROVED** [review]
It looks good! Just a few comments
- i've left on each of the files - like whether you'd still need them currently (since you've written them as 'for testing purpose only'. 
- I realize there are also timers used in multiple files like: `GameOverActions.java`, `MainGameScreen.java`, `PauseMenuActions.java`, `MainMenuActions.java`
Just like to comment that it will be difficult to test these later on with timers having to be created on a new thread (if im not mistaken). We've used timers before and had to switch to using the game time instead for unit testing. Might be worth exploring the possibility of using the game engine's time instead (if possible, if not could probably just do functional testing i suppose). 
- I've also added comments on possibly adding java doc (just a short explanation on why this file is needed and created for).
- There are also methods which have been labeled as deprecated - would this still be needed when merging into main? (just a minor thing - was just wondering)
- I've also looked at the prologue and epilogue - i thought they looked awesome and really fits the theme of this game. Great job! I notice that you can press ESC for prologue and epilogue. I got this error upon clicking ESC for epilogue
image

- story manager singleton is probably a good idea too! Background music to safehouse and level 3 sounds really good too! Music for game over screen was the best by far! Apart from that, i think its ok for a merge.

File: source/core/src/main/com/deco2800/game/ui/terminal/commands/StartTestCutSceneCommand.java (hunk 10 above)
Comment (implex001) [reply source/core/src/main/com/deco2800/game/ui/terminal/commands/StartTestCutSceneCommand.java:1]: This was used for testing purposes, and is still required as it provides a convenient trigger for a cutscene at any point in the game

**Pull request merged**